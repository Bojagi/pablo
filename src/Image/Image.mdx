---
name: Image
route: /image
menu: Components
---

import { Props, Playground } from 'docz'
import { Box } from '../Box/Box.tsx'
import { Image } from './Image.tsx'

# Image

The `Image` component is a plain proxy of the normal HTML `img` element but it adds support for the
styled-system layouting props (like margins, paddings, positioning, etc.).

## Examples

### Plain Image

You can display `IconButton`s in different sizes. By default they have the size `medium`. 
They can also be `small` and `large`. 

<Playground>
  <Image src="https://bojagi.io/images/bojagi-social-preview.jpg" width={500} />
</Playground>

### Layouting an image

You can use margin to give images a bit of space around them

<Playground>
  <Box bgColor="papayaWhip">
  <Image mx={2} my={4} src="https://bojagi.io/images/bojagi-social-preview.jpg" width={500} />
  </Box>
</Playground>

Or you can use positioning

<Playground>
  <Box position="relative" m={20} height={600} width={600} bgColor="papayaWhip">
  <Image position="absolute" top={50} left={200} src="https://bojagi.io/images/bojagi-social-preview.jpg" width={500} />
  </Box>
</Playground>

## Import

```js
import { Image } from '@bojagi/pablo/Image';
```

## Props

<Props of={Image} />
