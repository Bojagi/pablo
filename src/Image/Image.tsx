import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { LayoutBoxProps, layoutInterpolationFn } from '../Box';

export interface ImageProps
  extends Omit<React.ImgHTMLAttributes<HTMLImageElement>, 'color'>,
    LayoutBoxProps {
  className?: string;
  src: string;
  alt?: string;
  width?: number;
  height?: number;
}

const StyledImg = styled.img<LayoutBoxProps>`
  ${layoutInterpolationFn}
`;

export const Image = forwardRef<HTMLImageElement, ImageProps>((props, ref) => (
  <StyledImg ref={ref} {...props} />
));
