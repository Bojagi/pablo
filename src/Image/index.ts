// import './styles';

export * from './Image';
