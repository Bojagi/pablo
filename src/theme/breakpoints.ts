export type Breakpoint = 'sm' | 'md' | 'lg' | 'xl';

export const breakpointNames: Breakpoint[] = ['sm', 'md', 'lg', 'xl'];

export type BreakpointsArray = [number, number, number, number];

export type Breakpoints = BreakpointsArray &
  Record<Breakpoint, number> & {
    breakpointNames: Breakpoint[];
  };

export function createBreakpoints(
  array: BreakpointsArray,
  usedBreakpointNames: Breakpoint[] = breakpointNames
): Breakpoints {
  const newBreakpoints = ([...array] as unknown) as Breakpoints;
  usedBreakpointNames.forEach((bp, index) => {
    newBreakpoints[bp] = newBreakpoints[index];
  });
  newBreakpoints.breakpointNames = usedBreakpointNames;

  return newBreakpoints as Breakpoints;
}

export const defaultBreakpoints: BreakpointsArray = [600, 1000, 1200, 1920];

export const breakpoints = createBreakpoints(defaultBreakpoints, breakpointNames);
