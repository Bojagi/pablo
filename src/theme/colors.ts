export type ShadedColor = {
  main: string;
  dark: string;
  darkest: string;
  light: string;
  lightest: string;
  contrastText: string;
  contrastTextLight: string;
};

export type CommonColors = {
  white: string;
  black: string;
  whiteContrastText: string;
  blackContrastText: string;
};

export type TextColors = {
  main: string;
  info: string;
};

export type BorderColors = {
  main: string;
  lightest: string;
};

export interface GrayColors {
  '50': string;
  '100': string;
  '200': string;
  '300': string;
  '400': string;
  '500': string;
  '600': string;
  '700': string;
  '800': string;
  '900': string;
}

export type AllColors = ShadedColor & TextColors & BorderColors & CommonColors;

export interface Colors {
  common: CommonColors;
  gray: GrayColors;
  borders: BorderColors;
  brand: ShadedColor;
  positive: ShadedColor;
  negative: ShadedColor;
  neutral: ShadedColor;
  text: TextColors;
}

export const WHITE = '#ffffff';
export const BLACK = '#000000';

export const colors: Colors = {
  common: {
    white: WHITE,
    black: BLACK,
    whiteContrastText: BLACK,
    blackContrastText: WHITE,
  },
  gray: {
    '50': getGrayColor(0.95),
    '100': getGrayColor(0.9),
    '200': getGrayColor(0.8),
    '300': getGrayColor(0.7),
    '400': getGrayColor(0.6),
    '500': getGrayColor(0.5),
    '600': getGrayColor(0.4),
    '700': getGrayColor(0.3),
    '800': getGrayColor(0.2),
    '900': getGrayColor(0.1),
  },
  borders: {
    main: `${BLACK}${getHexByte(0.25)}`,
    lightest: `${BLACK}${getHexByte(0.1)}`,
  },
  text: {
    main: BLACK,
    info: `${BLACK}${getHexByte(0.5)}`,
  },
  brand: {
    lightest: '#EBE7FF',
    light: '#D6CEFF',
    main: '#6A50F2',
    dark: '#4129BD',
    darkest: '#281683',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
  positive: {
    lightest: '#DFF6E6',
    light: '#BBECCB',
    main: '#38C765',
    dark: '#2FB85A',
    darkest: '#108B36',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
  negative: {
    lightest: '#FFE7E5',
    light: '#FFCFCC',
    main: '#F21D0D',
    dark: '#CA271B',
    darkest: '#871A12',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
  neutral: {
    lightest: '#FCF5DB',
    light: '#FFEB9E',
    main: '#FFDD5B',
    dark: '#EDC52A',
    darkest: '#D6AC0B',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
};

/**
 * Get hex color byte (e.g. to get transparency)
 * @param value Percentage value (between 0.0 and 1.0)
 */
function getHexByte(value: number) {
  return (Math.round((255 * value) / 16) * 16).toString(16);
}

function getGrayColor(value: number) {
  return `#${getHexByte(value)}${getHexByte(value)}${getHexByte(value)}`;
}
