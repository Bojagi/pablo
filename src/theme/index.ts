export * from './defaultTheme';
export * from './context';
export * from './PabloThemeProvider';
export * from './useComponentStyle';
