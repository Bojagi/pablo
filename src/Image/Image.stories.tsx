import React from 'react';
import { Image } from './Image';

export default {
  title: 'Image',
};

export const WithoutSize = () => <Image src="https://bojagi.io/images/bojagi-social-preview.jpg" />;

export const WithWidth = () => (
  <Image src="https://bojagi.io/images/bojagi-social-preview.jpg" width={200} />
);

export const WithHeight = () => (
  <Image src="https://bojagi.io/images/bojagi-social-preview.jpg" height={200} />
);
