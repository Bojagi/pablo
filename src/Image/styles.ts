import { registerComponentStyles } from '../theme/defaultComponentStyles';
import { BaseStyles } from '../types';

export type ImageStyleProperties = 'root';

export interface ImageStyles extends BaseStyles<ImageStyleProperties> {}

export const imageStyles: ImageStyles = {};

registerComponentStyles('image', imageStyles);
