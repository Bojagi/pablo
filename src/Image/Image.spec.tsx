import { render } from '@testing-library/react';
import React from 'react';
import { PabloThemeProvider } from '../theme';
import { Image } from './Image';

test('Render image', () => {
  const { container } = renderComponent({
    alt: 'Social media preview',
    src: 'https://bojagi.io/images/bojagi-social-preview.jpg',
  });
  expect(container).toMatchSnapshot();
});

function renderComponent(props) {
  return render(
    <PabloThemeProvider>
      <Image {...props} />
    </PabloThemeProvider>
  );
}
