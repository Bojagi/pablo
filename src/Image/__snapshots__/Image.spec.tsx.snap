// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Render image 1`] = `
.emotion-0 {
  box-sizing: border-box;
}

<div>
  <img
    alt="Social media preview"
    class="emotion-0"
    src="https://bojagi.io/images/bojagi-social-preview.jpg"
  />
</div>
`;
